text/0000-stabilize-catch-panic.md

```diff
@@ -0,0 +1,345 @@
+call to `set_len` (the next 10 elements are uninitialized). If the type `T`'s
+`clone` method panics then this broken invariant will escape the function. The
+broken `Vec` is then observed during its destructor, leading to the eventual
+memory unsafety.
+
+It's important to keep in mind that panic safety in Rust is not solely limited
+to memory safety. *Logical invariants* are often just as critical to keep correct
+during execution and no `unsafe` code in Rust is needed to break a logical
+invariant. In practice, however, these sorts of bugs are rarely observed due to
+Rust's design:
+
+* Rust doesn't expose uninitialized memory
+* Panics cannot be caught in a thread
+* Across threads data is poisoned by default on panics
+* Idiomatic Rust must opt in to extra sharing across boundaries (e.g. `RefCell`)
+* Destructors are relatively rare and uninteresting in safe code
+
+These mitigations all address the *second* aspect of exception unsafety:
+observation of broken invariants. With the tactics in place, it ends up being
+the case that **safe Rust code can largely ignore exception safety
+concerns**. That being said, it does not mean that safe Rust code can *always*
+ignore exception safety issues. There are a number of methods to subvert the
+mitigation strategies listed above:
+
+1. When poisoning data across threads, antidotes are available to access
+   poisoned data. Namely the [`PoisonError` type][pet] allows safe access to the
+   poisoned information.
+2. Single-threaded types with interior mutability, such as `RefCell`, allow for
+   sharing data across stack frames such that a broken invariant could
+   eventually be observed.
+3. Whenever a thread panics, the destructors for its stack variables will be run
+   as the thread unwinds. Destructors may have access to data which was also
+   accessible lower on the stack (such as through `RefCell` or `Rc`) which has a
+   broken invariant, and the destructor may then witness this.
+
+[pet]: http://doc.rust-lang.org/std/sync/struct.PoisonError.html
+
+But all of these "subversions" fall outside the realm of normal, idiomatic, safe
+Rust code, and so they all serve as a "heads up" that panic safety might be an
+issue. Thus, in practice, Rust programmers worry about exception safety far less
+than in languages with full-blown exceptions.
+
+Despite these methods to subvert the mitigations placed by default in Rust, a
+key part of exception safety in Rust is that **safe code can never lead to
+memory unsafety**, regardless of whether it panics or not. Memory unsafety
+triggered as part of a panic can always be traced back to an `unsafe` block.
+
+With all that background out of the way now, let's take a look at the guts of
+this RFC.
+
+# Detailed design
+
+At its heart, the change this RFC is proposing is to move
+`std::thread::catch_panic` to a new `std::panic` module and rename the function
+to `catch`. Additionally, the `Send` and `'static` bounds from the closure
```

Comment:
While `panic::catch_panic` *is* a bit redundant, it's possible that we might want to add a `catch` keyword in connection with `Result`-handling at some point (e.g. #243), and although we could probably finesse the situation somehow (language versioning, deprecation, etc.), it might still be preferable to try to preemptively avoid stepping on our own possible-future toes.